Content.Server/Atmos/Consoles/AtmosMonitoringConsoleSystem.cs

```diff
@@ -0,0 +1,541 @@
+        allChunks[chunkOrigin] = chunk;
+        _gridAtmosPipeChunks[gridUid] = allChunks;
+
+        // Update the components of the monitoring consoles that are attached to the same grid
+        var query = AllEntityQuery<AtmosMonitoringConsoleComponent, TransformComponent>();
+
+        while (query.MoveNext(out var ent, out var entConsole, out var entXform))
+        {
+            if (gridUid != entXform.GridUid)
+                continue;
+
+            entConsole.AtmosPipeChunks = allChunks;
+            Dirty(ent, entConsole);
+        }
+    }
+
+    private void UpdateAtmosPipeChunk(EntityUid uid, NodeContainerComponent nodeContainer, AtmosPipeColorComponent pipeColor, int tileIdx, ref AtmosPipeChunk chunk)
+    {
+        // Entities that are actively being deleted are not to be drawn
+        var stage = MetaData(uid).EntityLifeStage;
+
+        if (MetaData(uid).EntityLifeStage >= EntityLifeStage.Terminating)
+            return;
+
+        foreach ((var id, var node) in nodeContainer.Nodes)
+        {
+            if (node is not PipeNode)
+                continue;
+
+            var pipeNode = (PipeNode)node;
+            var netId = GetPipeNodeNetId(pipeNode);
+            var pipeDirection = pipeNode.CurrentPipeDirection;
+
+            chunk.AtmosPipeData.TryGetValue((netId, pipeColor.Color.ToHex()), out var atmosPipeData);
+            atmosPipeData |= (ulong)pipeDirection << tileIdx * SharedNavMapSystem.Directions;
+            chunk.AtmosPipeData[(netId, pipeColor.Color.ToHex())] = atmosPipeData;
+        }
+    }
+
+    private bool TryGettingFirstPipeNode(EntityUid uid, [NotNullWhen(true)] out PipeNode? pipeNode, [NotNullWhen(true)] out int? netId)
+    {
+        pipeNode = null;
+        netId = null;
+
+        if (!TryComp<NodeContainerComponent>(uid, out var nodeContainer))
+            return false;
+
+        foreach (var node in nodeContainer.Nodes.Values)
+        {
+            if (node is PipeNode)
+            {
+                pipeNode = (PipeNode)node;
+                netId = GetPipeNodeNetId(pipeNode);
+
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private int GetPipeNodeNetId(PipeNode pipeNode)
+    {
+        if (pipeNode.NodeGroup is BaseNodeGroup)
+        {
+            var nodeGroup = (BaseNodeGroup)pipeNode.NodeGroup;
+
+            return nodeGroup.NetId;
+        }
+
+        return -1;
+    }
+
+    #endregion
+
+    #region Initialization functions
+
+    private void InitializeAtmosMonitoringConsole(EntityUid uid, AtmosMonitoringConsoleComponent component)
+    {
+        var xform = Transform(uid);
+
+        if (xform.GridUid == null)
+            return;
+
+        var grid = xform.GridUid.Value;
+
+        if (!TryComp<MapGridComponent>(grid, out var map))
+            return;
+
+        component.AtmosDevices = GetAllAtmosDeviceNavMapData(grid);
+
+        if (!_gridAtmosPipeChunks.ContainsKey(grid))
```

Comment:
Use `TryGetValue` to avoid having to search for the key a second time in the else clause.